DynamicArrayHW/DynamicArray.cpp

```diff
@@ -0,0 +1,147 @@
+#include "DynamicArray.h"
+#include <xpolymorphic_allocator.h>
+
+template<typename T>
+bool DynamicArray<T>::operator==(const DynamicArray& other) const
+{
+	if (size != other.size)
+	{
+		return false;
+	}
+
+	for (size_t i = 0; i < other.size; i++)
+	{
+		if (array[i] != other.array[i])
+		{
+			return false;
+		}
+	}
+
+	return true;
+}
+
+template<typename T>
+void DynamicArray<T>::shrinkToFit()
+{
+	if (size == 0)
+	{
+		delete[] array;
+		array = nullptr;
+		capacity = 0;
+	}
+	else
+	{
+		T* newArray = new T[size];
+		for (size_t i = 0; i < size; i++)
+		{
+			newArray[i] = array[i];
+		}
+
+		delete[] array;
+		array = newArray;
+		capacity = size;
+	}
+}
+
+template<typename T>
+std::size_t DynamicArray<T>::getCapacity() const
+{
+	return capacity;
+}
+
+template<typename T>
+T DynamicArray<T>::back() const
+{
+	if (size == 0)
+	{
+		return -1;
+	}
+
+	return array[this->getSize - 1];
+}
+
+template<typename T>
+void DynamicArray<T>::pop_back()
+{
+	if (size == 0)
+	{
+		return;
+	}
+
+	T* newArray = new T[size - 1];
+	for (size_t i = 0; i < size - 1; i++)
+	{
+		newArray[i] = array[i];
+	}
+
+	delete[] array;
+	array = newArray;
+
+	size--;
+}
+
+template<typename T>
+void DynamicArray<T>::clear()
+{
+	delete[] array;
+	size = 0;
+	capacity = 0;
+	array = nullptr;
+}
+
+template<typename T>
+std::size_t DynamicArray<T>::getSize() const
+{
+	return size;
+}
+
+template<typename T>
+void DynamicArray<T>::setSize(std::size_t newSize)
+{
+	std::size_t minSize = std::min(size, newSize);
+	T* newArray = new T[minSize];
+	for (std::size_t i = 0; i < minSize; i++)
+	{
+		newArray[i] = array[i];
+	}
+
+	delete[] array;
+	size = newSize;
+	if (capacity < size)
+	{
+		capacity = size;
+	}
+	array = newArray;
+
+	delete[] newArray;
+}
+
+template<typename T>
+T& DynamicArray<T>::operator[](std::size_t index)
+{
+	return array[index];
+}
+
+template<typename T>
+T& DynamicArray<T>::operator=(const T& other)
+{
+	if (this == &other) {
+		return *this;
+	}
+
+	delete[] array;
+
+	size = other.size;
+	if (capacity < size)
+	{
+		capacity = size;
+	}
+	array = new T[other.size];
+	for (size_t i = 0; i < other.size; i++)
+	{
+		array[i] = other.array[i];
+	}
+
+	return *this;
+}
```

Comment:
The compilation is not broken for as long as you don't use them. The moment you'll try it will break once again.
ALL template definitions should in header

Reply:
Plus, you should test all functionality of your class, not the easiest one to check or to implement only